### ISSUE TITLE
IP::getNonProxyIpFromHeader retrieves final proxy instead of client

### ISSUE THREAD

**Issue by jcracknell** [piwik/piwik#7060]
[`IP::getNonProxyIpFromHeader`](https://github.com/piwik/piwik/blob/638a84a09de3bfe2ddd02e563af1f39b1fa8b303/core/IP.php#L288) attempts to retrieve the client IP address from headers configured in `proxy_client_headers[]`. This calls `IP::getLastIpFromList`, excluding proxies configured via `proxy_ips[]`.

What I do not understand is why by default this returns the *last* IP, whereas the format for `X-Forwarded-For` is `client, proxy1, proxy2, ...`:
http://en.wikipedia.org/wiki/X-Forwarded-For#Format

This only becomes an issue when running Piwik behind multiple proxies; for example the configuration in question is:

    [Enterprise Appliance] => [IIS ARR] => [Piwik]

So Piwik sees:

    X-Forwarded-For: <client>, <enterprise_appliance>

Basically the current behavior would seem to select the IP of the last proxy by default. This would be problematic in a scenario with variable proxy IPs.

**Comment by tsteur** [comment]
Thx for the report!

**Comment by jcracknell** [comment]
I thought about this some more, and the current behavior of `proxy_ips[]` works something like [nginx's notion of trusted proxies][1] I ran into while flailing around with this issue. The problem is that it is not documented as such, and the behavior falls apart because it accepts an `X-Forwarded-For` address regardless of whether or not it was provided by a member of `proxy_ips[]`:

Consider a client issuing a bogus XFF header, in `client => proxy1 => proxy2 => Piwik`:

    [client => proxy1] X-Forwarded-For: bogus
    [proxy1 => proxy2] X-Forwarded-For: bogus, client
    [proxy2 =>  piwik] X-Forwarded-For: bogus, client, proxy1

I think the behavior should perhaps check if the direct client (`$_SERVER['REMOTE_ADDR']`) is in `proxy_ips[]` and only pull ips from `proxy_client_headers[]` if this is the case, taking the last non-proxy IP in the list. This lets you accept both proxied and unproxied requests.

The downside is this would make `proxy_ips[]` required configuration for all proxied installs. You could of course put this behavior behind a configuration flag, e.g. `proxy_trust_required`.

[1]: http://serverfault.com/questions/314574/nginx-real-ip-header-and-x-forwarded-for-seems-wrong

**Comment by djschoone** [comment]
I think you would not need the `proxy_trust_required` (e.d.) flag, because it's possible to check if the array of `proxy_ips` has values.

As seen in function `getNonProxyIpFromHeader()` in core/IP.php the `$default` (which is in most cases `$_SERVER['REMOTE_ADDR']` is added to the list of `proxy_ips` so that way it should never be returned by `getLastIpFromList`

How can I help, to get this issue resolved? I'm currently implementing Piwik behind multiple proxies so we are in the blind of the users real IP at this moment.

**Comment by mattab** [comment]
> How can I help, to get this issue resolved?

I didn't check the details of this issue, but it would help if you manage to create a Pull request and explain in this PR what you do, why, what problem it solves, etc. Cheers

**Comment by robocoder** [comment]
To clarify, this works as expected/designed.

Each successive proxy appends to the Forwarded-For header. As the OP observes:

```X-Forwarded-For: <client>, <enterprise_appliance>```

`proxie_ips` contains the IP for the enterprise_appliance, so, `IP::getLastIpFromList` filters it out, and gets the client IP. 

The reason why we traverse the list of IPs in `X-Forwarded-For` from right-to-left is because we don't automatically trust anything at the beginning of the header (e.g., it could be forged).

**Comment by mattab** [comment]
Thanks for the details @robocoder 
Therefore it seems we could close this issue with "Wontfix" / Works as expected.
Unless one can prove there is really a problem/limitation with the current logic?

**Comment by djschoone** [comment]
Thanks for the follow up. I cannot recall how we we're able to resolve the issue. We somehow managed it, probably by changing the request header done by the proxy server.

**Comment by mbrodala** [comment]
So this one can be closed I assume.

---

> I think the behavior should perhaps check if the direct client ($_SERVER['REMOTE_ADDR']) is in proxy_ips[] and only pull ips from proxy_client_headers[] if this is the case, taking the last non-proxy IP in the list. This lets you accept both proxied and unproxied requests.

This a very good thinking and deserves its own, separate ticket. Especially given the fact that calling the HTTP tracking API with `cip` requires a token but doing a request with `X-Forwarded-For` is accepted if `proxy_client_headers` is configured accordingly.

This is what TYPO3 does here BTW:

1. [Determine if the `REMOTE_ADDR` matches a configured proxy IP mask list](https://github.com/TYPO3/typo3/blob/f6b1178d4e8a25704223c3ebd1f0bf9e0f1faddb/typo3/sysext/core/Classes/Http/NormalizedParams.php#L737)
2. [Determine the remote IP address respecting proxy headers](https://github.com/TYPO3/typo3/blob/f6b1178d4e8a25704223c3ebd1f0bf9e0f1faddb/typo3/sysext/core/Classes/Http/NormalizedParams.php#L706)
